hera_cal/smooth_cal.py

```diff
@@ -82,55 +82,96 @@ def dpss_filters(freqs, times, freq_scale=10, time_scale=1800, eigenval_cutoff=1
         time_scale: time scale in seconds. Defined analogously to freq_scale.
             Note that time_scale is in seconds, times is in days.
         eigenval_cutoff: sinc_matrix eigenvalue cutoff to use for included dpss modes.
-            Only used when the filtering method is 'DPSS'
+        Nmax: Maximum number of time/frequency axis samples to attempt to compute the DPSS
+            vectors for. If the number of time/frequency samples exceeds this threshold value,
+            the number of non-zero eigenvalues will be estimated to attempt to reduce the computation
+            time of the DPSS vectors while still keeping all vectors whose eigenvalue is greater than
+            eigenval_cutoff. Default is 2000 grid points.
 
     Returns:
-        filters: DPSS filtering vectors, ndarray of size (Ntimes * Nfreqs, N_frequency_vectors * N_time_vectors)
+        time_filters: DPSS filtering vectors along the time axis, ndarray of size (Ntimes, N_time_vectors)
+        freq_filters: DPSS filtering vectors along the frequency axis, (Nfreqs, N_freq_vectors)
     """
     delay_scale = (freq_scale * 1e6) ** -1  # Puts it in seconds
     fringe_scale = (time_scale) ** -1  # fringe scale in Hz
     time_in_seconds = (times - times.min()) * 60 * 60 * 24  # time array in seconds
 
-    freq_filters, nfreq_comps = uvtools.dspec.dpss_operator(
-        freqs,
-        filter_centers=[0],
-        filter_half_widths=[delay_scale],
-        eigenval_cutoff=[eigenval_cutoff],
-    )  # DPSS filters along the frequency-axis
-    freq_filters = freq_filters.real.astype(np.float32)
-
-    time_filters, ntime_comps = uvtools.dspec.dpss_operator(
-        time_in_seconds,
-        filter_centers=[0],
-        filter_half_widths=[fringe_scale],
-        eigenval_cutoff=[eigenval_cutoff],
-    )  # DPSS filters along the time-axis
-    time_filters = time_filters.real.astype(np.float32)
-
-    filters = np.reshape(
-        time_filters[:, None, :, None] * freq_filters[None, :, None, :],
-        (times.shape[0] * freqs.shape[0], ntime_comps[0] * nfreq_comps[0]),
-    )  # 2D time-frequency DPSS vectors
-    return filters
-
-
-def fit_solution_matrix(weights, design_matrix):
+    dpss_windows = []
+
+    for x0, fw in zip([time_in_seconds, freqs], [fringe_scale, delay_scale]):
+
+        if Nmax < x0.shape[0]:
+            Nw = 2 * x0.shape[0] * fw * np.diff(x0)[0]
+            Nterms = int(min(10 * Nw, x0.shape[0]))
```

Comment:
Where does 10 come from?

Reply:
I meant to add that as an argument. I'll add more comments to this function, but the idea behind this speed-up is that when we generate these vectors, we generate the whole set (N vectors of size N) which takes a while to compute for large N.  Instead, what I'm attempting to do is estimate the number of vectors we need so that we don't have to compute all of the DPSS vectors, but just enough to where the minimum eigenvalue is at least as small as your eigenvalue cutoff. When the number of grid points in your DPSS vectors is large (N > 1000) you can estimate the number of DPSS vectors with close to unity eigenvalues using the approximation Nw = 2 * N * fw * dt. The `10 * Nw` was just providing a buffer so that we're sure to include enough vectors to meet the eigenvalue cut-off. Since the fall-off in the eigenvalues from 1 to 0 happens quickly, the value of `10` provided more than enough vectors for standard smoothcal inputs. There may be a better way to do this though

Reply:
I actually found a paper that gives an upper bound of the number of eigenvalues greater than some minimum eigenvalue so the buffer as an input argument won't be important.

Reply:
Feel free to cite the paper in the comments.

Reply:
I cited two of the papers I used in the comments